===================================================================
RCS file: /cvs/glibc/libc/stdio-common/vfscanf.c,v
retrieving revision 1.110.2.6
retrieving revision 1.110.2.7
diff -u -r1.110.2.6 -r1.110.2.7
--- libc/stdio-common/vfscanf.c	2007/05/04 10:05:57	1.110.2.6
+++ libc/stdio-common/vfscanf.c	2007/07/08 10:08:25	1.110.2.7
@@ -530,12 +530,17 @@
 	{
 	  /* Eat whitespace.  */
 	  int save_errno = errno;
-	  errno = 0;
+	  __set_errno (0);
 	  do
-	    if (__builtin_expect (inchar () == EOF && errno == EINTR, 0))
+	    /* We add the additional test for EOF here since otherwise
+	       inchar will restore the old errno value which might be
+	       EINTR but does not indicate an interrupt since nothing
+	       was read at this time.  */
+	    if (__builtin_expect ((c == EOF || inchar () == EOF)
+				  && errno == EINTR, 0))
 	      input_error ();
 	  while (ISSPACE (c));
-	  errno = save_errno;
+	  __set_errno (save_errno);
 	  ungetc (c, s);
 	  skip_space = 0;
 	}
